mail_tracking/controllers/main.py

```diff
@@ -4,41 +4,53 @@
 import werkzeug
 from psycopg2 import OperationalError
 from odoo import api, http, registry, SUPERUSER_ID
+from odoo.addons.mail.controllers.main import MailController
+from odoo.http import request
 import logging
 import base64
 _logger = logging.getLogger(__name__)
 
 BLANK = 'R0lGODlhAQABAIAAANvf7wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=='
 
 
-def _env_get(db, callback, tracking_id, event_type, **kw):
-    res = 'NOT FOUND'
-    reg = False
-    current = http.request.db and db == http.request.db
-    env = current and http.request.env
-    if not env:
-        with api.Environment.manage():
-            try:
-                reg = registry(db)
-            except OperationalError:
-                _logger.warning("Selected BD '%s' not found", db)
-            except Exception:  # pragma: no cover
-                _logger.warning("Selected BD '%s' connection error", db)
-            if reg:
-                _logger.info("New environment for database '%s'", db)
-                with reg.cursor() as new_cr:
-                    new_env = api.Environment(new_cr, SUPERUSER_ID, {})
-                    res = callback(new_env, tracking_id, event_type, **kw)
-                    new_env.cr.commit()
-    else:
-        # make sudo when reusing environment
-        env = env(user=SUPERUSER_ID)
-        res = callback(env, tracking_id, event_type, **kw)
-    return res
-
-
-class MailTrackingController(http.Controller):
+# Helper class to invoke "_tracking_*" methods
+class EnvironmentDB:
+    def __init__(self, db):
+        self.db = db
+        self.__is_new = False
+        self._env = False
 
+    def __enter__(self):
+        reg = False
+        current = http.request.db and self.db == http.request.db
+        self._env = current and http.request.env
+        if not self._env:
+            with api.Environment.manage():
+                try:
+                    reg = registry(self.db)
+                except OperationalError:
+                    _logger.warning("Selected BD '%s' not found", self.db)
+                except Exception:  # pragma: no cover
+                    _logger.warning(
+                        "Selected BD '%s' connection error", self.db)
+                if reg:
+                    self.__is_new = True
+                    _logger.info("New environment for database '%s'", self.db)
+                    with reg.cursor() as new_cr:
+                        self._env = api.Environment(new_cr, SUPERUSER_ID, {})
+                        return self._env
+        else:
+            self._env = self._env(user=SUPERUSER_ID)
+        return self._env
+
+    def __exit__(self, type, value, traceback):
+        if self.__is_new:
+            self._env.cr.commit()
+
+
+class MailTrackingController(MailController):
+
+    # Prepare remote info metadata
```

Comment:
Is there a reason why you put a comment over the method instead of a proper docstring in the method? It could be just the same text, but in the right place... :thinking: